src/array/data_chunk_builder.rs

```diff
@@ -0,0 +1,58 @@
+use std::iter::IntoIterator;
+
+use itertools::Itertools;
+
+use super::{ArrayBuilderImpl, DataChunk};
+use crate::types::{DataType, DataValue};
+
+pub struct DataChunkBuilder {
+    array_builders: Vec<ArrayBuilderImpl>,
+    size: usize,
+    capacity: usize,
+}
+
+impl DataChunkBuilder {
+    pub fn new<'a>(data_types: impl IntoIterator<Item = &'a DataType>, capacity: usize) -> Self {
+        assert_ne!(capacity, 0);
+        let array_builders = data_types
+            .into_iter()
+            .map(|ty| ArrayBuilderImpl::with_capacity(capacity, ty))
+            .collect();
+        DataChunkBuilder {
+            array_builders,
+            size: 0,
+            capacity,
+        }
+    }
+
+    /// Push a row in the Iterator.
+    ///
+    /// The row is accepted as an iterator of [`DataValue`], and it's required that the size of row
+    /// should be the same as the number of columns.
+    ///
+    /// Either a [`DataChunkBuilder`] or a [`DataChunk`] will be returned, depends on whether `size
+    /// == capacity`
+    pub fn push_row(mut self, row: impl IntoIterator<Item = DataValue>) -> Result<Self, DataChunk> {
```

Comment:
What if we need to append multi rows? Call this one by one may introduce many unnecessary mathod calls.

Reply:
Of course, we can introduce push_rows on demand, but it's not necessary now :(

Reply:
But it may affect signature.

Reply:
It will be another method

Reply:
maybe
```rust
    // panic if columns have different size
    pub fn push_multi_rows(mut self, columns: impl IntoIterator<Item = Array>) -> Result<Self, DataChunk> {
```